opm/simulators/wells/StandardWell_impl.hpp

```diff
@@ -186,25 +190,33 @@ namespace Opm
         // Note: currently, the WQTotal definition is still depends on Injector/Producer.
         assert(comp_idx < num_components_);
 
-        if (well_type_ == INJECTOR) { // only single phase injection
-            // TODO: using comp_frac here is dangerous, it should be changed later
-            // Most likely, it should be changed to use distr, or at least, we need to update comp_frac_ based on distr
-            // while solvent might complicate the situation
-            const auto pu = phaseUsage();
-            const int legacyCompIdx = ebosCompIdxToFlowCompIdx(comp_idx);
-            double comp_frac = 0.0;
-            if (has_solvent && comp_idx == contiSolventEqIdx) { // solvent
-                comp_frac = comp_frac_[pu.phase_pos[ Gas ]] * wsolvent();
-            } else if (legacyCompIdx == pu.phase_pos[ Gas ]) {
-                comp_frac = comp_frac_[legacyCompIdx];
-                if (has_solvent) {
-                    comp_frac *= (1.0 - wsolvent());
+        if (this->isInjector()) { // only single phase injection
+            double inj_frac = 0.0;
+            switch (this->wellEcl().injectorType()) {
+            case Well2::InjectorType::WATER:
+                if (comp_idx == FluidSystem::waterCompIdx) {
+                    inj_frac = 1.0;
                 }
-            } else {
-                comp_frac = comp_frac_[legacyCompIdx];
+                break;
+            case Well2::InjectorType::GAS:
+                if (has_solvent && comp_idx == contiSolventEqIdx) { // solvent
+                    inj_frac = wsolvent();
+                } else if (comp_idx == FluidSystem::gasCompIdx) {
+                    inj_frac = has_solvent ? 1.0 - wsolvent() : 1.0;
+                }
+                break;
+            case Well2::InjectorType::OIL:
+                if (comp_idx == FluidSystem::oilCompIdx) {
+                    inj_frac = 1.0;
+                }
+                break;
+            case Well2::InjectorType::MULTI:
+                // Not supported.
+                // deferred_logger.warning("MULTI_PHASE_INJECTOR_NOT_SUPPORTED",
+                //                         "Multi phase injectors are not supported, requested for well " + name());
+                break;
```

Comment:
Why not throw

Reply:
Cannot throw in single-well classes without catching in the calling method from BlackoilWellModel, that will lead to deadlocks.